cardano-api 1.36.0

Unisay opened the pull request:
This PR updates the `cardano-api` version used in the wallet to `1.36.0` to make use of the new `BundledProtocolParameters`.

Unisay committed:
cardano-api 1.36.0

jonathanknowles commented on lib/wallet/test/unit/Cardano/Wallet/Shelley/TransactionSpec.hs:
```diff
@@ -4111,9 +4111,9 @@ prop_balanceTransactionValid wallet (ShowBuildable partialTx) seed
+            not $ WriteTx.isBelowMinimumCoinForTxOut era ledgerPParams out
           where
-            pp = ledgerPParams
+
```
```suggestion
```

jonathanknowles commented on lib/wallet/test/unit/Cardano/Wallet/Shelley/TransactionSpec.hs:
```diff
@@ -4301,11 +4303,11 @@ hasReturnCollateral (Cardano.Tx (Cardano.TxBody content) _) =
-  where
-    nodePParams = Cardano.ProtocolParameters
+mockProtocolParametersForBalancing ::
+    (ProtocolParameters, Cardano.ProtocolParameters)
```
No need to reformat the type signature here, as it was actually already in our standard format, which is:

```suggestion
mockProtocolParametersForBalancing
    :: (ProtocolParameters, Cardano.ProtocolParameters)
```

(with `::` on a new line)
(See examples within the expandable section of https://github.com/input-output-hk/adrestia/blob/master/docs/code/Coding-Standards.md#avoid-variable-length-indentation)

Unisay replied:
One unfortunate side effect is that syntax highlighting doesn't recognize function name as such and doesn't highlight it when `::` is on the next line.

jonathanknowles commented on cabal.project:
```diff
@@ -161,8 +161,8 @@ constraints:
-  , cardano-api >= 1.35.4
-  , cardano-node >= 1.35.4
+  , cardano-api >= 1.35.6
+  , cardano-node >= 1.35.6
```
These bounds are just lower bounds.

What we actually get is:
```
$ cardano-node --version
cardano-node 1.36.0 - linux-x86_64 - ghc-8.10
git rev 84a871ed86fc1b86018231b9cb24ba7b83493c0d
```

We could just set this to:
```
  , cardano-api >= 1.36.0
  , cardano-node >= 1.36.0
```
Or
```
  , cardano-api  == 1.36.0
  , cardano-node == 1.36.0
```

Anviking commented on lib/wallet/src/Cardano/Wallet/Shelley/Transaction.hs:
```diff
@@ -1336,7 +1337,8 @@ assignScriptRedeemers pparams ti utxo redeemers tx =
         let pparams' = Cardano.toLedgerPParams
-                Cardano.ShelleyBasedEraAlonzo pparams
+                Cardano.ShelleyBasedEraAlonzo
+                    (Cardano.unbundleProtocolParams pparams)
```
Here we should call `unbundleLedgerShelleyBasedProtocolParams` instead of `toLedgerPParams . unbundleProtocolParams` to actually leverage the `BundledProtocolParameters` abstraction and avoid costly conversions 

=> https://github.com/input-output-hk/cardano-wallet/pull/3819

Anviking committed:
Remove all calls to Cardano.toLedgerPParams in balanceTx

The point of `BundledProtocolParameters era` is that it already contains
`Ledger.PParams ledgerEra`. Now we don't have to call the expensive
conversion `Cardano.toLedgerPParams`, but we can rather call the cheap
`unbundleLedgerShelleyBasedProtocolParams`.

Unisay committed:
chore: reformat

Anviking commented on lib/wallet/api/http/Cardano/Wallet/Api/Http/Shelley/Server.hs:
```diff
@@ -3071,12 +3071,14 @@ balanceTransaction
                         , "withRecentEra should prevent this being reached."
                         ])
+                    (Cardano.bundleProtocolParams
+                        (WriteTx.fromRecentEra (WriteTx.recentEra @era)))
```
We still have one remaining `Ledger.PParams -> Cardano.ProtocolParameters -> Ledger.PParams` roundtrip and the second step is here. 

https://github.com/input-output-hk/cardano-node/blob/a16c19850c7c54f72cb9b4ef45e1934f37309cb6/cardano-api/src/Cardano/Api/ProtocolParameters.hs#L1382

So completely replacing `Cardano.ProtocolParameters` with `Ledger.PParams` (ADP-2459) is still an opportunity for speedup which remains.

Anviking commented on lib/wallet/src/Cardano/Wallet.hs:
```diff
@@ -2253,12 +2253,14 @@ buildTransactionPure
                 , "withRecentEra should prevent this to be reached."
                 ])
+            (Cardano.bundleProtocolParams
+                (WriteTx.fromRecentEra (WriteTx.recentEra @era)))
```
Same applies here: https://github.com/input-output-hk/cardano-wallet/pull/3812/files#r1153379942

Anviking replied:
Actually, for fee estimation's sake, simply I guess just moving this conversion out of the `calculateFeePercentiles` loop would help 🤔

Anviking replied:
Don't see any effect here: https://github.com/input-output-hk/cardano-wallet/pull/3822
